x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshot.java

```diff
@@ -95,8 +94,17 @@ public SourceOnlySnapshot(Directory targetDirectory) {
                     newInfos.add(newInfo);
                 }
             }
+            if (existingSegmentInfos != null && newInfos.equals(existingSegmentInfos.asList())) {
+                return Collections.emptyList();
```

Comment:
I wonder if we need to do a deeper "equals" here, for example, whether we need to take `userData` into account as well (or even go deeper, requiring `lastGeneration` to be the same).

I would rather err on the cautious side here (i.e. take more factors into account).

Longer tem, we should try to leverage sequence numbers to detect that nothing needs to be uploaded (which also helps in scenarios where we had a primary failover on an index that's essentially read-only).

Reply:
Note that these `SegmentInfos` are created by us in the `_snapshot` directory. We are not comparing against data in the real index. So the `generation` as well as the `userData` on the commit are whatever we set it to here. Since we are manually setting the generation and not messing with the `userData` in this method, I think just comparing the individual segment info is safe?

Reply:
We can commit new user data on a shard without adding a new segment. While none of the userData committed substantively effects what's being restored currently, I wonder if this will introduce subtle bugs if the snapshot should reflect that change. I would prefer for this to be handled at a higher level (i.e. general snapshot functionality, determining whether to skip uploading any files if there are no effective changes to the snapshot, possibly based on sequence numbers).

Reply:
I thought about this a little now and I agree, this is a fair point. I think we'll need to have another mechanism to detect changes as you point out. I'm closing here and will deal with this in a more robust way in the overall fix for source only snapshots.
Thanks!